File: test/compare.test.ts
import assert from "node:assert/strict";
import test from "node:test";

import { compareReceipts } from "../src/compare.js";
import { sha256 } from "../src/hash.js";
import type { RunReceipt, TheseusContract } from "../src/types.js";

function receipt(profile: string, values: { exit: number; stdout: string; stderr?: string; runtime?: string }): RunReceipt {
  const stdout = values.stdout;
  const stderr = values.stderr ?? "";
  const runtimeCanonical = values.runtime ?? "runtime-a";
  return {
    subject: "installed",
    profile,
    installArgs: profile === "noOptional" ? ["--omit=optional"] : [],
    environment: { platform: "darwin", arch: "arm64", node: "v24", npm: "11" },
    runtime: {
      kind: "installed",
      packageName: "fixture",
      packageVersion: "1.0.0",
      binName: "fixture",
      executablePath: "node_modules/.bin/fixture",
      executableRealPath: "node_modules/fixture/bin.js",
      optionalDependencies: [],
      canonical: runtimeCanonical,
      sha256: sha256(runtimeCanonical),
    },
    probes: [
      {
        id: "help",
        argv: ["--help"],
        exitCode: values.exit,
        signal: null,
        timedOut: false,
        stdout,
        stderr,
        stdoutSha256: sha256(stdout),
        stderrSha256: sha256(stderr),
        durationMs: 1,
      },
    ],
  };
}

const baseContract: TheseusContract = {
  version: 1,
  profiles: {
    default: { installArgs: [] },
    noOptional: { installArgs: ["--omit=optional"] },
  },
  probes: [{ id: "help", argv: ["--help"], timeoutMs: 5_000, compare: ["exit", "stdout", "stderr", "runtime"] }],
  normalizers: [],
  allowedDeltas: [],
};

test("reports exit, output, and runtime drift independently", () => {
  const result = compareReceipts(baseContract, [
    receipt("default", { exit: 0, stdout: "full\n", runtime: "native" }),
    receipt("noOptional", { exit: 1, stdout: "fallback\n", runtime: "fallback" }),
  ]);
  assert.deepEqual(
    result.findings.map((finding) => finding.code),
    ["THS-PARITY-001", "THS-PARITY-002", "THS-RUNTIME-001"],
  );
});

test("an exact allowed delta does not suppress unrelated drift", () => {
  const defaultReceipt = receipt("default", { exit: 0, stdout: "full\n", stderr: "", runtime: "native" });
  const fallbackReceipt = receipt("noOptional", {
    exit: 0,
    stdout: "fallback\n",
    stderr: "unexpected\n",
    runtime: "native",
  });
  const contract: TheseusContract = {
    ...baseContract,
    allowedDeltas: [
      {
        probe: "help",
        between: ["default", "noOptional"],
        field: "stdout",
        fromSha256: sha256("full\n"),
        toSha256: sha256("fallback\n"),
        reason: "documented reduced help",
      },
    ],
  };
  const result = compareReceipts(contract, [defaultReceipt, fallbackReceipt]);
  assert.equal(result.allowedDifferences.length, 1);
  assert.deepEqual(result.findings.map((finding) => finding.field), ["stderr"]);
});

test("checks declared exit expectations per profile", () => {
  const contract: TheseusContract = {
    ...baseContract,
    probes: [{ ...baseContract.probes[0]!, expect: { exit: 0 } }],
  };
  const result = compareReceipts(contract, [
    receipt("default", { exit: 0, stdout: "same" }),
    receipt("noOptional", { exit: 2, stdout: "same" }),
  ]);
  assert.ok(result.findings.some((finding) => finding.code === "THS-EXPECT-001"));
});

test("checks typed stdout JSON paths", () => {
  const contract: TheseusContract = {
    ...baseContract,
    probes: [
      {
        ...baseContract.probes[0]!,
        compare: ["exit"],
        expect: {
          stdoutJson: [
            { path: "meta.version", type: "string" },
            { path: "meta.ready", type: "boolean" },
          ],
        },
      },
    ],
  };
  const pass = compareReceipts(contract, [receipt("default", { exit: 0, stdout: '{"meta":{"version":"1","ready":true}}' })]);
  assert.deepEqual(pass.findings, []);

  const fail = compareReceipts(contract, [receipt("default", { exit: 0, stdout: '{"meta":{"version":1}}' })]);
  assert.deepEqual(fail.findings.map((finding) => finding.code), ["THS-EXPECT-002", "THS-EXPECT-002"]);
});

File: internal/compare/compare_test.go
package compare

import (
	"testing"

	"github.com/JeremyDev87/theseus/internal/canonical"
	"github.com/JeremyDev87/theseus/internal/model"
)

func pointer(value int) *int { return &value }

func TestReportsIndependentExitOutputRuntimeDrift(t *testing.T) {
	leftOut, rightOut := "left\n", "right\n"
	contract := model.Contract{Probes: []model.ProbeConfig{{ID: "help", Compare: []model.CompareField{model.FieldExit, model.FieldStdout, model.FieldRuntime}}}}
	receipts := []model.RunReceipt{
		{Profile: "default", Runtime: model.RuntimeIdentity{SHA256: "runtime-a", Canonical: "a"}, Probes: []model.ProbeReceipt{{ID: "help", ExitCode: pointer(0), Stdout: leftOut, StdoutSHA256: canonical.SHA256String(leftOut)}}},
		{Profile: "noOptional", Runtime: model.RuntimeIdentity{SHA256: "runtime-b", Canonical: "b"}, Probes: []model.ProbeReceipt{{ID: "help", ExitCode: pointer(1), Stdout: rightOut, StdoutSHA256: canonical.SHA256String(rightOut)}}},
	}
	result := Receipts(contract, receipts)
	want := []string{"THS-PARITY-001", "THS-PARITY-002", "THS-RUNTIME-001"}
	if len(result.Findings) != len(want) {
		t.Fatalf("findings=%#v", result.Findings)
	}
	for index, code := range want {
		if result.Findings[index].Code != code {
			t.Fatalf("finding %d=%s want %s", index, result.Findings[index].Code, code)
		}
	}
}

func TestJSONLargeExponentRemainsANumber(t *testing.T) {
	contract := model.Contract{Probes: []model.ProbeConfig{{ID: "json", Expect: &model.ProbeExpectation{StdoutJSON: []model.JSONPathExpectation{{Path: "value", Type: "number"}}}}}}
	receipts := []model.RunReceipt{{Profile: "default", Probes: []model.ProbeReceipt{{ID: "json", Stdout: `{"value":1e400}`}}}}
	result := Receipts(contract, receipts)
	if len(result.Findings) != 0 {
		t.Fatalf("unexpected findings %#v", result.Findings)
	}
}

func TestJSONPointerAndOwnProperties(t *testing.T) {
	contract := model.Contract{Probes: []model.ProbeConfig{{ID: "json", Expect: &model.ProbeExpectation{StdoutJSON: []model.JSONPathExpectation{{Path: "/meta/a.b", Type: "number"}, {Path: "/items/0/name", Type: "string"}, {Path: "/items/length", Type: "number"}, {Path: "/items/00/name", Type: "string"}, {Path: "/__proto__/polluted", Type: "string"}}}}}}
	receipts := []model.RunReceipt{{Profile: "default", Probes: []model.ProbeReceipt{{ID: "json", Stdout: `{"meta":{"a.b":1},"items":[{"name":"ok"}]}`}}}}
	result := Receipts(contract, receipts)
	if len(result.Findings) != 2 || result.Findings[0].Actual != "undefined" || result.Findings[1].Actual != "undefined" {
		t.Fatalf("unexpected findings %#v", result.Findings)
	}
}

func TestExactAllowedDeltaDoesNotSuppressUnrelatedDrift(t *testing.T) {
	leftOut, rightOut := "full\n", "fallback\n"
	leftErr, rightErr := "", "unexpected\n"
	contract := model.Contract{
		Probes: []model.ProbeConfig{{ID: "help", Compare: []model.CompareField{model.FieldStdout, model.FieldStderr}}},
		AllowedDeltas: []model.AllowedDelta{{
			Probe: "help", Between: [2]string{"default", "noOptional"}, Field: model.FieldStdout,
			FromSHA256: canonical.SHA256String(leftOut), ToSHA256: canonical.SHA256String(rightOut), Reason: "documented reduced help",
		}},
	}
	receipts := []model.RunReceipt{
		comparisonReceipt("default", 0, leftOut, leftErr),
		comparisonReceipt("noOptional", 0, rightOut, rightErr),
	}
	result := Receipts(contract, receipts)
	if len(result.AllowedDifferences) != 1 || len(result.Findings) != 1 || result.Findings[0].Field != "stderr" {
		t.Fatalf("unexpected exact-delta result %#v", result)
	}
}

func TestExitAndTypedJSONExpectationsArePerProfile(t *testing.T) {
	zero := 0
	contract := model.Contract{Probes: []model.ProbeConfig{{
		ID: "json", Compare: []model.CompareField{model.FieldExit},
		Expect: &model.ProbeExpectation{Exit: &zero, StdoutJSON: []model.JSONPathExpectation{{Path: "meta.version", Type: "string"}, {Path: "/escaped~1key/~0value", Type: "boolean"}}},
	}}}
	receipts := []model.RunReceipt{
		comparisonReceipt("default", 0, `{"meta":{"version":"1"},"escaped/key":{"~value":true}}`, ""),
		comparisonReceipt("noOptional", 2, `{"meta":{"version":1},"escaped/key":{}}`, ""),
	}
	result := Receipts(contract, receipts)
	want := []string{"THS-EXPECT-001", "THS-EXPECT-002", "THS-EXPECT-002", "THS-PARITY-001"}
	if len(result.Findings) != len(want) {
		t.Fatalf("findings=%#v", result.Findings)
	}
	for index, code := range want {
		if result.Findings[index].Code != code {
			t.Fatalf("finding %d=%s want %s", index, result.Findings[index].Code, code)
		}
	}
}

func comparisonReceipt(profile string, exit int, stdout, stderr string) model.RunReceipt {
	return model.RunReceipt{
		Profile: profile,
		Runtime: model.RuntimeIdentity{SHA256: "same-runtime", Canonical: "same-runtime"},
		Probes: []model.ProbeReceipt{{
			ID: "help", ExitCode: pointer(exit), Stdout: stdout, Stderr: stderr,
			StdoutSHA256: canonical.SHA256String(stdout), StderrSHA256: canonical.SHA256String(stderr),
		}, {
			ID: "json", ExitCode: pointer(exit), Stdout: stdout, Stderr: stderr,
			StdoutSHA256: canonical.SHA256String(stdout), StderrSHA256: canonical.SHA256String(stderr),
		}},
	}
}
